crates/accelerate/src/circuit_library/entanglement.rs

```diff
@@ -0,0 +1,197 @@
+/// that in each layer (controlled by ``offset``) the entanglement gates are shifted by one, plus
+/// in each second layer, the entanglement gate is turned upside down.
+/// Offset 0 -> [(2,3,0), (3,0,1), (0,1,2), (1,2,3)]
+/// Offset 1 -> [(3,2,1), (0,3,2), (1,0,3), (2,1,0)]
+/// Offset 2 -> [(0,1,2), (1,2,3), (2,3,0), (3,0,1)]
+/// ...
+pub fn shift_circular_alternating(
+    num_qubits: u32,
+    block_size: u32,
+    offset: usize,
+) -> Box<dyn Iterator<Item = Vec<u32>>> {
+    // index at which we split the circular iterator -- we use Python-like indexing here,
+    // and define ``split`` as equivalent to a Python index of ``-offset``
+    let split = (num_qubits - (offset as u32 % num_qubits)) % num_qubits;
+    let shifted = circular(num_qubits, block_size)
+        .skip(split as usize)
+        .chain(circular(num_qubits, block_size).take(split as usize));
+    if offset % 2 == 0 {
+        Box::new(shifted)
+    } else {
+        // if the offset is odd, reverse the indices inside the qubit block (e.g. turn CX
+        // gates upside down)
+        Box::new(shifted.map(|indices| indices.into_iter().rev().collect()))
+    }
+}
+
+/// Get an entangler map for an arbitrary number of qubits.
+///
+/// Args:
+///     num_qubits: The number of qubits of the circuit.
+///     block_size: The number of qubits of the entangling block.
+///     entanglement: The entanglement strategy as string.
+///     offset: The block offset, can be used if the entanglements differ per block,
+///         for example used in the "sca" mode.
+///
+/// Returns:
+///     The entangler map using mode ``entanglement`` to scatter a block of ``block_size``
+///     qubits on ``num_qubits`` qubits.
+pub fn get_entanglement_from_str(
+    num_qubits: u32,
+    block_size: u32,
+    entanglement: &str,
+    offset: usize,
+) -> PyResult<Box<dyn Iterator<Item = Vec<u32>>>> {
+    if block_size > num_qubits {
+        return Err(QiskitError::new_err(format!(
+            "block_size ({}) cannot be larger than number of qubits ({})",
+            block_size, num_qubits
+        )));
+    }
+
+    if entanglement == "pairwise" && block_size > 2 {
+        return Err(QiskitError::new_err(format!(
+            "block_size ({}) can be at most 2 for pairwise entanglement",
+            block_size
+        )));
+    }
+
+    // if block size is 2 and pairwise, this is just linear: [0, 1, 2, ...]
+    match (entanglement, block_size) {
+        ("full", _) => Ok(Box::new(full(num_qubits, block_size))),
+        ("linear", _) => Ok(Box::new(linear(num_qubits, block_size))),
+        ("reverse_linear", _) => Ok(Box::new(reverse_linear(num_qubits, block_size))),
+        ("sca", _) => Ok(shift_circular_alternating(num_qubits, block_size, offset)),
+        ("circular", _) => Ok(circular(num_qubits, block_size)),
+        ("pairwise", 1) => Ok(Box::new(linear(num_qubits, 1))),
+        ("pairwise", 2) => Ok(Box::new(pairwise(num_qubits))),
```

Comment:
Oh, I see, you have simplified the original Python logic, or rather moved it to the individual functions. I think the code became much more clear now. However, do we have enough tests to cover all the corner-case scenarios: such as `block_size` being 1, `block_size` being `num_qubits`, etc.?

Reply:
I can add some more 🙂